File: main/3d-surface-area/3d_surface_area.rb
#!/usr/bin/env ruby
$VERBOSE = true

def read_record
  gets.split.map!(&:to_i)
end

def read_padded_row(width)
  row = read_record
  raise 'wrong row width' if row.size != width
  row.unshift(0).push(0).freeze
end

def read_padded_matrix(height, width)
  margin = Array.new(width + 2, 0).freeze
  rows = (0...height).map { read_padded_row(width) }
  rows.unshift(margin).push(margin).freeze
end

height, width = read_record
grid = read_padded_matrix(height, width)

area = height * width * 2 # the top and bottom faces

(1...grid.size).each do |i|
  (1...grid[i].size).each do |j|
    area += (grid[i][j - 1] - grid[i][j]).abs +
            (grid[i - 1][j] - grid[i][j]).abs
  end
end

puts area

File: main/3d-surface-area/3d-surface-area.cs
using System;
using System.Collections.Generic;
using System.Linq;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

internal static class Solution {
    private static string[] ReadTokens(int count)
    {
        var line = Console.ReadLine();

        var tokens = line.Split((char[])null,
                                StringSplitOptions.RemoveEmptyEntries);

        if (tokens.Length != count) {
            throw new ArgumentException(message: "wrong record length",
                                        paramName: nameof(count));
        }

        return tokens;
    }

    private static IEnumerable<int> ReadValues(int count)
        => ReadTokens(count).Select(int.Parse);

    private static int[,] ReadPaddedMatrix(int height, int width)
    {
        var grid = new int[height + 2, width + 2];

        for (var i = 1; i <= height; ++i) {
            using (var en = ReadValues(width).GetEnumerator()) {
                for (var j = 1; j <= width; ++j) {
                    en.MoveNext();
                    grid[i, j] = en.Current;
                }
            }
        }

        return grid;
    }

    private static void Main()
    {
        var dimensions = ReadValues(2).ToList();
        var height = dimensions[0];
        var width = dimensions[1];

        var grid = ReadPaddedMatrix(height, width);

        var area = height * width * 2; // top and bottom faces

        for (var i = 0; i <= height; ++i) {
            for (var j = 0; j <= width; ++j) {
                area += Math.Abs(grid[i, j] - grid[i, j + 1])
                      + Math.Abs(grid[i, j] - grid[i + 1, j]);
            }
        }

        Console.WriteLine(area);
    }
}
